flake.nix

```diff
@@ -29,9 +29,9 @@
 
       lib = {
         hm = import ./modules/lib { lib = nixpkgs.lib; };
-        homeManagerConfiguration = { configuration, system, homeDirectory
-          , username, extraModules ? [ ], extraSpecialArgs ? { }, pkgs
-          , lib ? pkgs.lib, check ? true, stateVersion ? "20.09" }@args:
+        homeManagerConfiguration = { configuration, pkgs, lib ? pkgs.lib
+          , homeDirectory, username, stateVersion, extraModules ? [ ]
```

Comment:
Do you know why `stateVersion` is here at all? Same with `username` and `homeDirectory`. Seems they could easily be defined inside `configuration`?

Reply:
I was wondering that as well. The problem if we remove them is that `stateVersion` becomes optional as it has a default value of 18.09, and this in turn seems to entail default values for `username` and `homeDirectory`, which we don't want.

I think `stateVersion` should either not have a default value, or default to `pkgs.lib.trivial.release` [like in nixpkgs](https://github.com/NixOS/nixpkgs/blob/e0a42267f73ea52adc061a64650fddc59906fc99/nixos/modules/misc/version.nix#L83-L85). Then maybe we can remove those arguments.

Reply:
Additionally if we remove the `stateVersion` argument we lose the ability to assert that it's ≥ 20.09, which is needed to avoid the uninformative error about looking up `<nixpkgs>` in pure mode. Maybe that's fine though, it's 2022.

Reply:
cc @berbiche @terlar as you might have thoughts on this

Reply:
As the person who added `stateVersion` I would be fine with removing it, but perhaps we should pass some option to the configuration letting Home Manager know the config came from a flake (or that it needs to be "pure") so we could assert the stateVersion is ≥ 20.09.

The error is quite confusing if you have an older `stateVersion` and use flakes.

There has been some discussion here as well #3041 2647

Reply:
I did try adding a module assertion directly in the flake, but it was useless as the problematic `<nixpkgs>` is evaluated first. Not sure if we can fix that.

Reply:
I've been meaning to remove the default value of `stateVersion` from the option definition and will do that in a separate commit. Will do that change before merging this.

Reply:
I've done that already: [`96ebe57` (#3037)](https://github.com/nix-community/home-manager/pull/3037/commits/96ebe57b47900d49e9989bc99973f991b83ca0e1)